### PR TITLE
Improve error message when CHPL_MEM=cstdlib is used with a registered heap

### PR DESCRIPTION
**Pull request by ronawho**
Improve the error message when CHPL_MEM=cstdlib is used in a configuration that
requires a registered heap. Also add some docs to chplenv.rst:CHPL_MEM to
provide more info about the error message.

Closes https://github.com/chapel-lang/chapel/issues/5649

### CODE REVIEW

**Comment by ronawho** [comment]
@gbtitus mind taking a look?

File: doc/rst/usingchapel/chplenv.rst [hunk 1]
```diff
@@ -385,6 +385,13 @@ CHPL_MEM
    If unset, ``CHPL_MEM`` defaults to ``jemalloc`` for most configurations.
    If the target platform is ``cygwin*`` it defaults to ``cstdlib``
 
+   .. note::
+     Certain ``CHPL_COMM`` settings (e.g. ugni and gasnet segment fast/large)
+     register the heap to improve communication performance.  Registering the
+     heap requires special allocator support that not all allocators provide.
+     Currently ``jemalloc`` is required for configurations that require a
```
Comment (gbtitus) [inline doc/rst/usingchapel/chplenv.rst]: Department of redundancy department (2 "required"s).  Maybe: "Currently only ``jemalloc`` is capable of supporting configurations that require a registered heap."

File: runtime/src/mem/cstdlib/mem-cstdlib.c [hunk 2]
```diff
@@ -35,8 +35,11 @@ void chpl_mem_layerInit(void) {
   size_t size;
 
   chpl_comm_desired_shared_heap(&start, &size);
-  if (start || size)
-    chpl_error("set CHPL_MEM to a more appropriate mem type", 0, 0);
+  if (start || size) {
```
Comment (gbtitus) [inline runtime/src/mem/cstdlib/mem-cstdlib.c]: Teensy nit: probably okay to just check `start`; in other configs we assert that start and size are either both zero or both nonzero.

**Comment by gbtitus** [comment]
Looks good modulo a couple of line comments (one definitely a nit) and could be merged as is; no need to look again if you change anything in response.

File: doc/rst/usingchapel/chplenv.rst (hunk 1 above)
Comment (ronawho) [inline doc/rst/usingchapel/chplenv.rst]: Changing to your suggested wording

**Pull request merged**